File: python/problem_33.py
import bisect
from typing import List


def running_median(nums: List[int]) -> List[float]:
    result = []
    sorted_so_far = []
    for v in nums:
        bisect.insort(sorted_so_far, v)
        n = len(sorted_so_far)
        if n % 2 == 1:
            result.append(float(sorted_so_far[n // 2]))
        else:
            result.append((sorted_so_far[n // 2 - 1] + sorted_so_far[n // 2]) / 2.0)
    return result


def main():
    print("Problem 33:")
    assert running_median([2, 1, 5, 7, 2, 0, 5]) == [2.0, 1.5, 2.0, 3.5, 2.0, 2.0, 2.0]
    assert running_median([1]) == [1.0]
    assert running_median([1, 2]) == [1.0, 1.5]
    print("All Tests Passed")

File: rust/problem_33.rs
pub fn running_median(nums: &[i64]) -> Vec<f64> {
    let mut sorted: Vec<i64> = Vec::new();
    let mut result = Vec::new();
    for &v in nums {
        let pos = sorted.partition_point(|&x| x < v);
        sorted.insert(pos, v);
        let n = sorted.len();
        let median = if n % 2 == 1 {
            sorted[n / 2] as f64
        } else {
            (sorted[n / 2 - 1] + sorted[n / 2]) as f64 / 2.0
        };
        result.push(median);
    }
    result
}

pub fn problem33() {
    println!("Problem 33:");
    let got = running_median(&[2,1,5,7,2,0,5]);
    let exp = vec![2.0, 1.5, 2.0, 3.5, 2.0, 2.0, 2.0];
    assert_eq!(got, exp);
    println!("All Tests Passed");
}
